File: mobile/shared/src/commonMain/kotlin/uk/foodclassifier/image/NormalizedRect.kt
package uk.foodclassifier.image

data class NormalizedRect(
    val left: Float,
    val top: Float,
    val right: Float,
    val bottom: Float,
) {
    init {
        require(left in 0f..1f && top in 0f..1f && right in 0f..1f && bottom in 0f..1f)
        require(right > left && bottom > top)
    }

    fun translated(normalizedDeltaX: Float, normalizedDeltaY: Float): NormalizedRect {
        val width = right - left
        val height = bottom - top
        var newLeft = left + normalizedDeltaX
        var newTop = top + normalizedDeltaY
        newLeft = newLeft.coerceIn(0f, 1f - width)
        newTop = newTop.coerceIn(0f, 1f - height)
        return NormalizedRect(
            left = newLeft,
            top = newTop,
            right = newLeft + width,
            bottom = newTop + height,
        )
    }

    companion object {
        /** Wide horizontal band — matches a typical ingredients block on product labels. */
        val ingredientsGuide = NormalizedRect(
            left = 0.04f,
            top = 0.36f,
            right = 0.96f,
            bottom = 0.58f,
        )

        val fullImage = NormalizedRect(
            left = 0f,
            top = 0f,
            right = 1f,
            bottom = 1f,
        )
    }
}

internal fun fitImageBounds(
    containerWidth: Float,
    containerHeight: Float,
    imageWidth: Float,
    imageHeight: Float,
): ImageBounds {
    if (containerWidth <= 0f || containerHeight <= 0f || imageWidth <= 0f || imageHeight <= 0f) {
        return ImageBounds(0f, 0f, containerWidth, containerHeight)
    }

    val imageAspect = imageWidth / imageHeight
    val containerAspect = containerWidth / containerHeight
    return if (imageAspect > containerAspect) {
        val displayedWidth = containerWidth
        val displayedHeight = containerWidth / imageAspect
        val yOffset = (containerHeight - displayedHeight) / 2f
        ImageBounds(0f, yOffset, displayedWidth, displayedHeight)
    } else {
        val displayedHeight = containerHeight
        val displayedWidth = containerHeight * imageAspect
        val xOffset = (containerWidth - displayedWidth) / 2f
        ImageBounds(xOffset, 0f, displayedWidth, displayedHeight)
    }
}

internal data class ImageBounds(
    val left: Float,
    val top: Float,
    val width: Float,
    val height: Float,
) {
    fun toScreenRect(normalized: NormalizedRect): ScreenRect = ScreenRect(
        left = left + normalized.left * width,
        top = top + normalized.top * height,
        right = left + normalized.right * width,
        bottom = top + normalized.bottom * height,
    )
}

fun mapViewGuideToImageGuide(
    guide: NormalizedRect,
    viewWidth: Float,
    viewHeight: Float,
    imageWidth: Float,
    imageHeight: Float,
): NormalizedRect {
    if (viewWidth <= 0f || viewHeight <= 0f || imageWidth <= 0f || imageHeight <= 0f) {
        return guide
    }

    val displayed = fitImageBounds(viewWidth, viewHeight, imageWidth, imageHeight)

    val guideLeftPx = guide.left * viewWidth
    val guideTopPx = guide.top * viewHeight
    val guideRightPx = guide.right * viewWidth
    val guideBottomPx = guide.bottom * viewHeight

    val contentLeft = displayed.left
    val contentTop = displayed.top
    val contentRight = displayed.left + displayed.width
    val contentBottom = displayed.top + displayed.height

    val clippedLeft = maxOf(guideLeftPx, contentLeft)
    val clippedTop = maxOf(guideTopPx, contentTop)
    val clippedRight = minOf(guideRightPx, contentRight)
    val clippedBottom = minOf(guideBottomPx, contentBottom)

    if (clippedRight <= clippedLeft || clippedBottom <= clippedTop) {
        return guide
    }

    return NormalizedRect(
        left = (clippedLeft - contentLeft) / displayed.width,
        top = (clippedTop - contentTop) / displayed.height,
        right = (clippedRight - contentLeft) / displayed.width,
        bottom = (clippedBottom - contentTop) / displayed.height,
    )
}

internal data class ScreenRect(
    val left: Float,
    val top: Float,
    val right: Float,
    val bottom: Float,
)

File: flutter/lib/image/normalized_rect.dart
/// Normalized rectangle (0–1) for crop guides and image cropping.
class NormalizedRect {
  const NormalizedRect({
    required this.left,
    required this.top,
    required this.right,
    required this.bottom,
  })  : assert(left >= 0 && left <= 1),
        assert(top >= 0 && top <= 1),
        assert(right >= 0 && right <= 1),
        assert(bottom >= 0 && bottom <= 1),
        assert(right > left),
        assert(bottom > top);

  final double left;
  final double top;
  final double right;
  final double bottom;

  double get width => right - left;
  double get height => bottom - top;

  /// Wide horizontal band — matches a typical ingredients block on product labels.
  static const ingredientsGuide = NormalizedRect(
    left: 0.04,
    top: 0.36,
    right: 0.96,
    bottom: 0.58,
  );

  static const fullImage = NormalizedRect(
    left: 0,
    top: 0,
    right: 1,
    bottom: 1,
  );

  NormalizedRect translated({
    required double normalizedDeltaX,
    required double normalizedDeltaY,
  }) {
    var newLeft = left + normalizedDeltaX;
    var newTop = top + normalizedDeltaY;
    newLeft = newLeft.clamp(0.0, 1.0 - width);
    newTop = newTop.clamp(0.0, 1.0 - height);
    return NormalizedRect(
      left: newLeft,
      top: newTop,
      right: newLeft + width,
      bottom: newTop + height,
    );
  }
}

class ImageBounds {
  const ImageBounds({
    required this.left,
    required this.top,
    required this.width,
    required this.height,
  });

  final double left;
  final double top;
  final double width;
  final double height;

  ScreenRect toScreenRect(NormalizedRect normalized) => ScreenRect(
        left: left + normalized.left * width,
        top: top + normalized.top * height,
        right: left + normalized.right * width,
        bottom: top + normalized.bottom * height,
      );
}

class ScreenRect {
  const ScreenRect({
    required this.left,
    required this.top,
    required this.right,
    required this.bottom,
  });

  final double left;
  final double top;
  final double right;
  final double bottom;
}

ImageBounds fitImageBounds({
  required double containerWidth,
  required double containerHeight,
  required double imageWidth,
  required double imageHeight,
}) {
  if (containerWidth <= 0 ||
      containerHeight <= 0 ||
      imageWidth <= 0 ||
      imageHeight <= 0) {
    return ImageBounds(
      left: 0,
      top: 0,
      width: containerWidth,
      height: containerHeight,
    );
  }

  final imageAspect = imageWidth / imageHeight;
  final containerAspect = containerWidth / containerHeight;
  if (imageAspect > containerAspect) {
    final displayedWidth = containerWidth;
    final displayedHeight = containerWidth / imageAspect;
    final yOffset = (containerHeight - displayedHeight) / 2;
    return ImageBounds(
      left: 0,
      top: yOffset,
      width: displayedWidth,
      height: displayedHeight,
    );
  }

  final displayedHeight = containerHeight;
  final displayedWidth = containerHeight * imageAspect;
  final xOffset = (containerWidth - displayedWidth) / 2;
  return ImageBounds(
    left: xOffset,
    top: 0,
    width: displayedWidth,
    height: displayedHeight,
  );
}

NormalizedRect mapViewGuideToImageGuide({
  required NormalizedRect guide,
  required double viewWidth,
  required double viewHeight,
  required double imageWidth,
  required double imageHeight,
}) {
  if (viewWidth <= 0 ||
      viewHeight <= 0 ||
      imageWidth <= 0 ||
      imageHeight <= 0) {
    return guide;
  }

  final displayed = fitImageBounds(
    containerWidth: viewWidth,
    containerHeight: viewHeight,
    imageWidth: imageWidth,
    imageHeight: imageHeight,
  );

  final guideLeftPx = guide.left * viewWidth;
  final guideTopPx = guide.top * viewHeight;
  final guideRightPx = guide.right * viewWidth;
  final guideBottomPx = guide.bottom * viewHeight;

  final contentLeft = displayed.left;
  final contentTop = displayed.top;
  final contentRight = displayed.left + displayed.width;
  final contentBottom = displayed.top + displayed.height;

  final clippedLeft = guideLeftPx > contentLeft ? guideLeftPx : contentLeft;
  final clippedTop = guideTopPx > contentTop ? guideTopPx : contentTop;
  final clippedRight =
      guideRightPx < contentRight ? guideRightPx : contentRight;
  final clippedBottom =
      guideBottomPx < contentBottom ? guideBottomPx : contentBottom;

  if (clippedRight <= clippedLeft || clippedBottom <= clippedTop) {
    return guide;
  }

  return NormalizedRect(
    left: (clippedLeft - contentLeft) / displayed.width,
    top: (clippedTop - contentTop) / displayed.height,
    right: (clippedRight - contentLeft) / displayed.width,
    bottom: (clippedBottom - contentTop) / displayed.height,
  );
}
